### PR TITLE
Shopping: Products JSON Schema Validation 

### PR DESCRIPTION
**Pull request by timarney**
## Context

In support of #11210 we need make the WoocCommerce plugin available with some sample products to query against to ensure a valid JSON schema is returned.

> Using this JSON schema for shoppable products, we can verify that the data we return from WooCommerce & Shopify matches the expected format. So most likely this is something for the PHPUnit integration tests.

## Summary

- Adds  WoocCommerce  plugin installation step to start script
- Adds sample product import

## To-do
- e2e test is failing which is expected as the value for product price isn't matching the schema

## Testing Instructions

- `npm run env:start`
-  note CLI outputs that WooCommerce is installed
-  note in the WP Admin - WooCommerce has imported sample products (via the sample xml file)

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Run `npm run env:start` --- this should install WooCommerce + sample products
2. Activate the WooCommerce plugin --- ensure sample products exists

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Partially addresses #11210

### CODE REVIEW

**Comment by timarney** [comment]
Noting this was failing on WP 5.5 -- not sure if that's a plugin issue or something else happening.  For now I have wrapped a WP latest check around the install.

https://github.com/GoogleForCreators/web-stories-wp/runs/6061324126?check_suite_focus=true

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+describe('Shopping', () => {
+  describe('Shopping schema', () => {
+    let stopRequestInterception;
+    withExperimentalFeatures(['shoppingIntegration']);
+
+
+    beforeAll(async () => {
+      await page.setRequestInterception(true);
+      stopRequestInterception = addRequestInterception((request) => {
+        // ... grab story json to validate against schema ?
```
Comment (timarney) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js:36]: @swissspidy @spacedmonkey 

Is there a call / endpoint I can intercept here to get the story config to compare against ?

I can see I'm getting data (via the story dev tools) after I run this e2e test

```
"type": "product",
          "product": {
            "productId": "wc-36",
            "productTitle": "T-Shirt with Logo",
            "productBrand": "",
            "productPrice": "18",
            "productPriceCurrency": "USD",
            "productImages": [],
            "aggregateRating": {
              "ratingValue": "0",
              "reviewCount": 0,
              "reviewUrl": "http://localhost:8899/product/t-shirt-with-logo"
            },
            "productDetails": "This is a simple product.",
            "productUrl": "http://localhost:8899/product/t-shirt-with-logo"
          },
```

File: packages/e2e-tests/src/specs/editor/shopping/schema.json [hunk 2]
```diff
@@ -0,0 +1,281 @@
+{
```
Comment (timarney) [inline packages/e2e-tests/src/specs/editor/shopping/schema.json:1]: We'll need to pull this in but adding for now to get a test working

File: includes/REST_API/Products_Controller.php [hunk 3]
```diff
@@ -383,11 +383,11 @@ static function( $image_id ) {
 				'productId'            => 'wc-' . $product->get_id(),
 				'productTitle'         => $product->get_title(),
 				'productBrand'         => '', // TODO: Figure out how to best provide that.
-				'productPrice'         => $product->get_price(),
+				'productPrice'         => number_format( (float) $product->get_price(), 2 ),
```
Comment (timarney) [inline includes/REST_API/Products_Controller.php]: This was failing the schema as it was coming back from WooCommerce as a string

File: includes/REST_API/Products_Controller.php [hunk 4]
```diff
@@ -383,11 +383,11 @@ static function( $image_id ) {
...
 				'productPriceCurrency' => get_woocommerce_currency(),
 				'productImages'        => $images,
 				'aggregateRating'      => [
-					'ratingValue' => $product->get_average_rating(),
+					'ratingValue' => number_format( (float) $product->get_average_rating(), 2 ),
```
Comment (timarney) [inline includes/REST_API/Products_Controller.php]: This was failing the schema as it was coming back from WooCommerce as a string

File: includes/REST_API/Products_Controller.php [hunk 5]
```diff
@@ -387,7 +387,7 @@ static function( $image_id ) {
 				'productPriceCurrency' => get_woocommerce_currency(),
 				'productImages'        => $images,
 				'aggregateRating'      => [
-					'ratingValue' => $product->get_average_rating(),
+					'ratingValue' => (int) $product->get_average_rating(),
```
Comment (swissspidy) [inline includes/REST_API/Products_Controller.php]: This looks incorrect to me.

The schema says `ratingValue` is a `number`, [which can be a floating point number](https://json-schema.org/understanding-json-schema/reference/numeric.html).

`get_average_rating()` returns a float.

This is important because the average rating could be something like 4.4. Don't want to cast this to an int and turn this into `4`.

File: bin/local-env/install-wordpress.sh [hunk 6]
```diff
@@ -139,6 +139,15 @@ fi
 echo -e $(status_message "Installing and activating RTL Tester plugin...")
 wp plugin install rtl-tester --activate --force --quiet
 
+echo -e $(status_message "Installing WordPress importer...")
+wp plugin install wordpress-importer --activate --force --quiet
+
+# Only install woocommerce on latest version of WordPress.
```
Comment (swissspidy) [inline bin/local-env/install-wordpress.sh:145]: Sounds reasonable; WooCommerce has a high WP version requirement (5.7+)

File: bin/local-env/install-wordpress.sh [hunk 7]
```diff
@@ -219,3 +228,11 @@ wp option patch insert web_stories_experiments enableSVG 0
...
 wp post list --post_type=attachment --format=yaml
 wp plugin list --format=yaml
+
+# Only install woocommerce on latest version of WordPress.
+if [ "$WP_VERSION" == "latest" ]; then
+	echo -e $(status_message "Import sample woocommerce products...")
+	wp import /var/www/html/wp-content/plugins/woocommerce/sample-data/sample_products.xml --authors=skip --quiet
+	# deactivate test etc... can activate as needed
+	wp plugin deactivate woocommerce 
+fi
```
Comment (swissspidy) [inline bin/local-env/install-wordpress.sh]: Note: missing newline at end of file

File: bin/local-env/install-wordpress.sh [hunk 8]
```diff
@@ -219,3 +228,11 @@ wp option patch insert web_stories_experiments enableSVG 0
 wp user list --format=yaml
 wp post list --post_type=attachment --format=yaml
 wp plugin list --format=yaml
+
+# Only install woocommerce on latest version of WordPress.
+if [ "$WP_VERSION" == "latest" ]; then
+	echo -e $(status_message "Import sample woocommerce products...")
+	wp import /var/www/html/wp-content/plugins/woocommerce/sample-data/sample_products.xml --authors=skip --quiet
+	# deactivate test etc... can activate as needed
+	wp plugin deactivate woocommerce
```
Comment (swissspidy) [inline bin/local-env/install-wordpress.sh:237]: Does the plugin need to be activated further above in the first place? Is it required for the import to work?

Otherwise we can remove this line & the `--activate` flag above.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+
+// eslint-disable-next-line import/no-extraneous-dependencies -- in package.json
+import { matchers } from 'jest-json-schema';
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: The dependency needs to be in the dependencies of `e2e-tests/package.json`, not the main `package.json`.

Then this lint warning will go away.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  takeSnapshot,
+} from '@web-stories-wp/e2e-test-utils';
+
+/**
+ * Internal dependencies
+ */
+import * as schema from './schema.json';
+
+// eslint-disable-next-line jest/require-hook -- matchers only for this set of tests
+expect.extend(matchers);
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: Let's move this to `bootstrap.js`. Should be fine to make this matcher available for all e2e tests

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    afterAll(async () => {
+      if ('latest' === WPVersion) {
+        await deactivatePlugin('woocommerce');
+      }
+    });
+
+    if (WPVersion !== 'latest') {
+      //eslint-disable-next-line jest/expect-expect, jest/no-focused-tests
+      it.only('minimum WordPress requirement not met', () => {});
+    }
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: Can use `minWPVersionRequired` from `e2e-utils` package.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+          const { type, product } = element;
+          //eslint-disable-next-line jest/no-conditional-in-test -- only add products
+          if (type === 'product') {
+            products.push(product);
+          }
+        });
+      });
+
+      expect(products).toHaveLength(1);
+      expect(products[0]).toMatchSchema(schema);
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: The product element data does not necessarily map to the product schema 1:1.

What's important is that the final product output when previewing a story is valid.

So instead of opening devtools, let's preview the story, grab the product config from the amp-story-shopping config and validate that one.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+      await takeSnapshot(page, 'insert product');
+      await publishStory();
+
+      // open story dev tools
+      await page.keyboard.down('Meta');
+      await page.keyboard.down('Alt');
+      await page.keyboard.down('Shift');
+      await page.keyboard.down('J');
+
+      await takeSnapshot(page, 'shopping schema');
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: Don't think we need a screenshot of the dev tools; doesn't sound useful.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    it('should should match a valid schema', async () => {
+      await createNewStory();
+      await expect(page).toClick('[aria-controls="library-pane-shopping"]');
+      await expect(page).toClick('[aria-label="Product"]');
+      await page.keyboard.press('ArrowDown');
+      await page.keyboard.press('Enter');
+      await expect(page).toClick('button', { text: 'Insert product' });
+      await takeSnapshot(page, 'insert product');
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: I think we can introduce this snapshot later once we add e2e tests for product insertion in general, not for the schema test.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+      it.only('minimum WordPress requirement not met', () => {});
+    }
+
+    it('should should match a valid schema', async () => {
+      await createNewStory();
+      await expect(page).toClick('[aria-controls="library-pane-shopping"]');
+      await expect(page).toClick('[aria-label="Product"]');
+      await page.keyboard.press('ArrowDown');
+      await page.keyboard.press('Enter');
+      await expect(page).toClick('button', { text: 'Insert product' });
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: It would be nice to know which product this is inserting. Maybe in the dropdown we can select a few specific products by title?

File: includes/REST_API/Products_Controller.php (hunk 5 above)
Comment (spacedmonkey) [reply includes/REST_API/Products_Controller.php]:
```suggestion
					'ratingValue' => (float) $product->get_average_rating(),
```

File: packages/e2e-tests/src/specs/editor/shopping/schema.json (hunk 2 above)
Comment (swissspidy) [reply packages/e2e-tests/src/specs/editor/shopping/schema.json:1]: Happy to help with pulling this in automatically, should be trivial

File: bin/local-env/install-wordpress.sh (hunk 8 above)
Comment (timarney) [reply bin/local-env/install-wordpress.sh:237]: Does the plugin need to be activated further above in the first place?

```bash
> STATUS: Import sample woocommerce products...
Failed to import &#8220;V-Neck T-Shirt&#8221;: Invalid post type product<br />Failed to import 
```
☝️ 

Without activating above the import fails

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js (hunk 9 above)
Comment (timarney) [reply packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: moved https://github.com/GoogleForCreators/web-stories-wp/pull/11250/commits/93cf9536bf636c5e58b44bb9463e5c0217eb590e

File: packages/e2e-tests/src/specs/editor/shopping/schema.json (hunk 2 above)
Comment (swissspidy) [reply packages/e2e-tests/src/specs/editor/shopping/schema.json:1]: Added a GitHub Actions workflow for this now. That should do the trick.

Comment (timarney) [reply packages/e2e-tests/src/specs/editor/shopping/schema.json:1]: Did you commit to this PR?  If yes think I rebased over it --- sorry.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js (hunk 12 above)
Comment (timarney) [reply packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: Updated to pull from the preview

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js (hunk 15 above)
Comment (timarney) [reply packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: updated to type in a product name

File: packages/e2e-tests/src/specs/editor/shopping/schema.json (hunk 2 above)
Comment (swissspidy) [reply packages/e2e-tests/src/specs/editor/shopping/schema.json:1]: yes... let me try to add it again

Comment (swissspidy) [reply packages/e2e-tests/src/specs/editor/shopping/schema.json:1]: My fault for pushing to your PR right away, but to prevent overrides form happening again in the future using `--force-with-lease` instead of `--force` is useful. (or merging instead of rebasing & force-pushing)

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js (hunk 11 above)
Comment (timarney) [reply packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: So ... I swapped this back to using minWPVersionRequired

The WP 5.5. e2e test run is failing as it's looking for a Plugin that doesn't exist --- any way we can avoid this outside of how I was doing it prior.


WP 5.5
```
No node found for selector: tr[data-slug="woocommerce"] .activate a

      29 |
      30 |   if (!disableLink) {
    > 31 |     await page.click(`tr[data-slug="${slug}"] .activate a`);
         |     ^
      32 |     await page.waitForSelector(`tr[data-slug="${slug}"] .deactivate a`);
      33 |   }
```

**Comment by timarney** [comment]
https://github.com/GoogleForCreators/web-stories-wp/runs/6119493501?check_suite_focus=true#step:9:14

@swissspidy  I could be wrong but I think we're running into the `AMP sanitization` you mentioned in the original PR 

https://github.com/GoogleForCreators/web-stories-wp/pull/11158/#issuecomment-1088570370


Screen Shot 2022-04-21 at 5 19 04 PM

**Comment by timarney** [comment]
> Note: need to disable AMP sanitization (e.g. by commenting out the relevant line in Output_Buffer) for now since this is not stable yet.

Re tested locally with the AMP sanitization off 

The e2e test is catching / working as expected.

Screen Shot 2022-04-22 at 6 07 01 AM

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        JSON.parse(
+          document.querySelector('amp-story-shopping-attachment script')
+            .textContent
+        )
+      );
+
+      await page.bringToFront();
+      await previewPage.close();
+      expect(data.items).toHaveLength(1);
+      expect(data.items[0]).toMatchSchema(schema);
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: Let's add a few more products at the beginning of the test and then loop through all `data.items` to do the schema validation on every product

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+      await page.type('[aria-label="Product"]', 'Hoodie with Zipper');
+      await page.keyboard.press('ArrowDown');
+      await page.keyboard.press('Enter');
+      await expect(page).toClick('button', { text: 'Insert product' });
+      await publishStory();
+      const previewPage = await previewStory(page);
+      await previewPage.waitForSelector('amp-story-shopping-attachment script');
+
+      // pull product data from the DOM
+      const data = await previewPage.evaluate(() =>
```
Comment (swissspidy) [inline packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: A `expect(data.items).toMatchInlineSnapshot()` call here might be useful, makes it easy to see at first glance what products we're expecting to be there.

**Comment by swissspidy** [comment]
I think updating the Composer dependencies helped with the AMP sanitization part.

For the failing preview I need to take a closer look on Monday. Is it working for you locally with `npm run test:e2e:interactive packages/e2e-tests/src/specs/editor/shopping/woocommerce.js`?

**Comment by timarney** [comment]
> Composer dependencies

Okay that would make sense

Locally it's still getting stuck 

Screen Shot 2022-04-22 at 6 03 36 PM


Screen Shot 2022-04-22 at 6 04 09 PM

**Comment by timarney** [comment]
Locally running the single test with Output buffering commented out it finishes the test.


Screen Shot 2022-04-22 at 6 15 27 PM
t

**Comment by swissspidy** [comment]
Hmm alright, then I didn‘t look close enough. Will re-check on Monday.

**Comment by timarney** [comment]
@swissspidy with your composer update commit merged back to this one --- this PR is now working as expected (fails on the schema check string vs number)

**Comment by swissspidy** [comment]
> @swissspidy I tested locally with your composer update commit merged back to this one it's s now working as expected (fails on the schema check string vs number)

Okay and which schema check is exactly failing? Can we fix our code?

I only see this error message:

```
  ● Shopping › Shopping schema › should should match a valid schema

    TimeoutError: waiting for selector `amp-story-shopping-attachment script` failed: timeout 30000ms exceeded
```

https://github.com/GoogleForCreators/web-stories-wp/runs/6174672630?check_suite_focus=true

**Comment by timarney** [comment]
> > @swissspidy I tested locally with your composer update commit merged back to this one it's s now working as expected (fails on the schema check string vs number)
> 
> Okay and which schema check is exactly failing? Can we fix our code?
> 
> I only see this error message:
> 
> ```
>   ● Shopping › Shopping schema › should should match a valid schema
> 
>     TimeoutError: waiting for selector `amp-story-shopping-attachment script` failed: timeout 30000ms exceeded
> ```
> 
> https://github.com/GoogleForCreators/web-stories-wp/runs/6174672630?check_suite_focus=true


Right for CI --- need to investigate why it's not finding that selector

Locally for me I'm getting a schema error because product price is returning a string vs the schema is looking for a number.  So yes that part can be fixed --- but maybe that fix should be in the product query PR?

**Comment by swissspidy** [comment]
> So yes that part can be fixed --- but maybe that fix should be in the product query PR?

Works for me; just keep in mind that then the test needs to be disabled until that gets fixed.

File: includes/REST_API/Products_Controller.php [hunk 18]
```diff
@@ -314,7 +314,8 @@ protected function get_items_shopify( $request ) {
 				'productBrand'         => $product['vendor'],
 				// TODO: Maybe eventually provide full price range.
 				// See https://github.com/ampproject/amphtml/issues/37957.
-				'productPrice'         => $product['priceRange']['minVariantPrice']['amount'],
+				// TODO: Replace hardcoded value with real price as a number (AMP schema validation).
+				'productPrice'         => 10,
```
Comment (swissspidy) [inline includes/REST_API/Products_Controller.php]: Why this change instead of type casting or something?

Comment (timarney) [reply includes/REST_API/Products_Controller.php]: Testing ...  if this passes will fix that up.

Comment (timarney) [reply includes/REST_API/Products_Controller.php]: Successful run here https://github.com/GoogleForCreators/web-stories-wp/actions/runs/2229993362

Updated to cast as a `float` for now.

File: packages/e2e-tests/src/specs/editor/shopping/woocommerce.js (hunk 17 above)
Comment (timarney) [reply packages/e2e-tests/src/specs/editor/shopping/woocommerce.js]: Added a snapshot here:

https://github.com/GoogleForCreators/web-stories-wp/pull/11250/commits/78a28b1f962742927056321fea8e7392a8f9fcad

**Comment by swissspidy** [comment]
All e2e tests seem to be passing now! 🎉

**Review by swissspidy — APPROVED** [review]

**Comment by swissspidy** [comment]
@spacedmonkey ping 👋  to review the latest version of this

File: .github/workflows/update-product-schema.yml [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+on:
+  workflow_dispatch:
+  schedule:
+    - cron: '0 12 * * 1'
+
+env:
+  GIT_AUTHOR_EMAIL: 94923726+googleforcreators-bot@users.noreply.github.com
+  GIT_AUTHOR_NAME: googleforcreators-bot
+  GIT_COMMITTER_EMAIL: 94923726+googleforcreators-bot@users.noreply.github.com
+  GIT_COMMITTER_NAME: googleforcreators-bot
```
Comment (spacedmonkey) [inline .github/workflows/update-product-schema.yml:12]: Should these be hardcoded or should they be setup as part of the project's env vars?

Comment (swissspidy) [reply .github/workflows/update-product-schema.yml:12]: Great question. They are hardcoded like this in all of our GitHub Actions. This is just copy-pasted from other workflows. Don't currently see the need for moving these to env vars, but maybe I can do that in the future if we feel like it would make things easier.
